lib/vnet/osconfig.go

```diff
@@ -0,0 +1,194 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"log/slog"
+	"net"
+	"os"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/vnet/v1"
+	"github.com/gravitational/teleport/api/profile"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/client"
+)
+
+type osConfig struct {
+	tunName    string
+	tunIPv4    string
+	tunIPv6    string
+	cidrRanges []string
+	dnsAddr    string
+	dnsZones   []string
+}
+
+type osConfigurator struct {
+	// Static config values.
+	tunName string
+	tunIPv6 string
+	dnsAddr string
+
+	// Computed config values.
+	tunIPv4 string
+
+	// Other state.
+	homePath           string
+	clientStore        *client.Store
+	clusterConfigCache *clusterConfigCache
+}
+
+func newOSConfigurator(ctx context.Context, tunName, ipv6Prefix, dnsAddr string) (*osConfigurator, error) {
+	homePath := os.Getenv(types.HomeEnvVar)
+	if homePath == "" {
+		// This runs as root so we need to be configured with the user's home path.
+		return nil, trace.BadParameter("%s must be set", types.HomeEnvVar)
+	}
+
+	// ipv6Prefix always looks like "fdxx:xxxx:xxxx::"
+	// Set the IPv6 address for the TUN to "fdxx:xxxx:xxxx::1", the first valid address in the range.
+	tunIPv6 := ipv6Prefix + "1"
+
+	configurator := &osConfigurator{
+		tunName:     tunName,
+		tunIPv6:     tunIPv6,
+		dnsAddr:     dnsAddr,
+		homePath:    homePath,
+		clientStore: client.NewFSClientStore(homePath),
+	}
+
+	clusterConfigCache, err := newClusterConfigCache(configurator.getVnetConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	configurator.clusterConfigCache = clusterConfigCache
+
+	return configurator, nil
+}
+
+func (c *osConfigurator) updateOSConfiguration(ctx context.Context) error {
+	var dnsZones []string
+	var cidrRanges []string
+
+	profileNames, err := profile.ListProfileNames(c.homePath)
+	if err != nil {
+		return trace.Wrap(err, "listing user profiles")
+	}
+	for _, profileName := range profileNames {
+		// TODO(nklaassen): support leaf clusters
+		vnetConfig, err := c.clusterConfigCache.getVnetConfig(ctx, profileName, "" /*leafClusterName*/)
+		if err != nil {
+			slog.WarnContext(ctx,
+				"Failed to load VNet configuration, profile may be expired, not configuring VNet for this cluster",
+				"profile", profileName, "error", err)
+			continue
+		}
```

Comment:
This is tricky. In both tsh and Connect, it's possible to start VNet while all your profiles are expired. This results in no resolvers being set up in `/etc/resolver`. If you try to connect to any of your apps under the DNS zone of the proxy service, those connections will not go through VNet and you won't get prompted to log in.

We could create resolvers for the proxy service DNS zone of all clusters, even if they are expired, but this won't work with custom DNS zones.

There's probably a way to make it smarter, e.g. leave around resolvers with backup nameservers even when turning off VNet, then cleaning them up on startup (e.g. if we notice that the user has logged out of the cluster, we remove the resolver; if we notice that the custom DNS zone is no longer defined on the backend, we remove the resolver). But idk if we have enough time before the release to tackle this.

It could be good to leave a good enough implementation for the release. E.g. always leave proxy service resolvers with nameserver backups and clean them up on launch. But I don't know how feasible it is to work on this right now, given that it might break other things.

Reply:
backup nameservers are not great, i wouldn't want to leave them around when not running vnet, it will add latency to every dns lookup.

If you start vnet without being logged in to a cluster I think it's okay to not handle requests for that cluster until you log in to it. `tsh vnet` I'm really not worried about, Connect should be the main supported way to run this, with the daemon in the future. If you're trying to connect to an app and it's not working, the first thing i'd do is open that cluster in connect and see if the app is there, you'll see pretty quickly that you're not logged in, then you'll log in and it will start working